pkg/v1/partial/index.go

```diff
@@ -0,0 +1,54 @@
+// Copyright 2020 Google LLC All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//    http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package partial
+
+import (
+	"bytes"
+	"fmt"
+
+	v1 "github.com/google/go-containerregistry/pkg/v1"
+)
+
+// WithRawIndexManifest defines the subset of v1.ImageIndex used by ImageByArch.
+type WithRawIndexManifest interface {
+	RawManifest() ([]byte, error)
+	Image(v1.Hash) (v1.Image, error)
+}
+
+// IndexManifest is a helper for implementing v1.ImageIndex.
+func IndexManifest(w WithRawIndexManifest) (*v1.IndexManifest, error) {
+	b, err := w.RawManifest()
+	if err != nil {
+		return nil, err
+	}
+	return v1.ParseIndexManifest(bytes.NewReader(b))
+}
+
+// ImageByArch returns the v1.Image belonging to the v1.ImageIndex with the
+// matching OS and Architecture.
+func ImageByArch(w WithRawIndexManifest, os, arch string) (v1.Image, error) {
+	im, err := IndexManifest(w)
+	if err != nil {
+		return nil, err
+	}
+	for _, manifest := range im.Manifests {
+		if manifest.Platform != nil &&
```

Comment:
I'd like to consolidate this logic if we can. There's already this: https://github.com/google/go-containerregistry/blob/919148f07b7c222ede2b9cf228e946b66a955808/pkg/v1/remote/index.go#L193-L221

I'm also not 100% sure that the existing logic there is sound (haven't read the spec carefully), but at the very least we are missing variant, which our platform flags allow you to specify.

Reply:
How about `ImageByPlatform` that takes a [`v1.Platform`](https://godoc.org/github.com/google/go-containerregistry/pkg/v1#Platform) which must match?